handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/available/AvailableProductMovesEndpoint.scala

```diff
@@ -44,44 +53,149 @@ object AvailableProductMovesEndpoint {
+      creditCardExpirationYear <- wireDefaultPaymentMethod.creditCardExpirationYear.toRight(())
+    } yield LocalDate.of(creditCardExpirationYear, creditCardExpirationMonth, 1)
+
+  extension[R, E, A] (zio: ZIO[R, E, A])
+    def handleError(message: String) = zio.catchAll {
+      error =>
+        ZIO.log(s"$message failed with: $error").flatMap(_ => ZIO.fail(InternalServerError))
+    }
+
+  private[productmove] def runWithEnvironment(subscriptionName: String): URIO[GetSubscription with GetCatalogue with GetAccount with Stage, OutputBody] = {
+    val output = for {
+      stage <- ZIO.service[Stage]
+      monthlyContributionRatePlanId = if (stage == Stage.DEV) "2c92c0f85a6b134e015a7fcd9f0c7855" else "2c92a0fc5aacfadd015ad24db4ff5e97"
+
       _ <- ZIO.log("subscription name: " + subscriptionName)
-      sub <- GetSubscription.get(if (subscriptionName == "true") "A-S00090478" else "A-S00339056") //DEV - for testing locally
-      _ <- ZIO.log("Sub: " + sub.toString)
-    } yield Success(List(
-      MoveToProduct(
-        id = "123",
-        name = "Digital Pack",
-        billing = Billing(
-          amount = Some(1199),
-          percentage = None,
-          currency = Currency.GBP,
-          frequency = Some(TimePeriod(TimeUnit.month, 1)),
-          startDate = Some("2022-09-21")
-        ),
-        trial = Some(Trial(14)),
-        introOffer = Some(Offer(
-          Billing(
-            amount = None,
-            percentage = Some(50),
-            currency = Currency.GBP,//FIXME doesn't make sense for a percentage
-            frequency = None,//FIXME doesn't make sense for a percentage
-            startDate = Some("2022-09-21")
-          ),
-          duration = TimePeriod(TimeUnit.month, 3)
-        ))
-      )
-    ))
+
+      // Kick off catalogue fetch in parallel
+      zuoraProductCatalogueFetch <- GetCatalogue.get.fork
+      subscription <- GetSubscription.get(subscriptionName).handleError("GetSubscription")
+
+      firstEligibilityChecks <-
+        (for {
+          _ <- succeedIfEligible(subscription.ratePlans.length == 1, s"Subscription: $subscriptionName has more than one ratePlan")
+          _ <- succeedIfEligible(subscription.ratePlans.head.productRatePlanId == monthlyContributionRatePlanId, s"Subscription: $subscriptionName is not a monthly contribution")
+          _ <- succeedIfEligible(subscription.ratePlans.head.ratePlanCharges.length == 1, s"Subscription: $subscriptionName has more than one ratePlan charge for ratePlan ${subscription.ratePlans.head}")
+        } yield ()).isSuccess
+
+      _ <- if (firstEligibilityChecks) ZIO.succeed(()) else ZIO.fail(Success(List()))
+
+      // Next payment date
+      chargedThroughDate <- ZIO.fromOption(subscription.ratePlans.head.ratePlanCharges.head.chargedThroughDate).orElse {
+        ZIO.log(s"chargedThroughDate is null for subscription $subscriptionName.").flatMap(_ => ZIO.fail(Success(List())))
+      }
```

Comment:
rather than using `.head` in many places, which might case an Element not found exception or pick the wrong plan if there are multiple ones, we could like this
```suggestion
      ratePlan <- asSingle(subscription.ratePlans, "ratePlan")
      charge <- asSingle(ratePlan.ratePlanCharges, "ratePlanCharge")
      chargedThroughDate <- ZIO.fromOption(charge.chargedThroughDate).orElseFail(s"chargedThroughDate is null")
      }
```
and define this function (copied from my uncommitted changes)
```scala
  def asSingle[A](list: List[A], message: String): IO[String, A] =
    list match {
      case singlePlan :: Nil => ZIO.succeed(singlePlan)
      case wrongNumber => ZIO.fail(s"subscription can't be cancelled as we didn't have a single $message: ${wrongNumber.length}: $wrongNumber")
    }
```
Then it will fail with a nice error if there is some assumption that doesn't hold.

Reply:
So for this we're checking there is just one ratePlan and one ratePlanCharge further up in the `subscriptionIsEligible` checks